bundles/org.eclipse.ui.editors/src/org/eclipse/ui/internal/texteditor/stickyscroll/StickyLinesProvider.java

```diff
@@ -0,0 +1,174 @@
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     SAP SE - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.ui.internal.texteditor.stickyscroll;
+
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+
+import org.eclipse.swt.custom.StyledText;
+
+import org.eclipse.jface.text.ITextViewerExtension5;
+import org.eclipse.jface.text.source.ISourceViewer;
+
+/**
+ * This class provides sticky lines for the given source code in the source viewer. The
+ * implementation is completely based on indentation and therefore should work by default for
+ * several languages.
+ */
+public class StickyLinesProvider {
+
+	private final static int IGNORE_INDENTATION= -1;
+
+	private final static String TAB= "\t"; //$NON-NLS-1$
+
+	private int tabWidth= 4;
+
+	/**
+	 * Calculate the sticky lines for the given source code in the source viewer for the given
+	 * vertical offset.
+	 * 
+	 * @param verticalOffset The vertical offset line index of the first visible line
+	 * @param sourceViewer The source viewer containing the source code
+	 * @return A list of sticky lines
+	 */
+	public List<StickyLine> get(int verticalOffset, ISourceViewer sourceViewer) {
+		LinkedList<StickyLine> stickyLines= new LinkedList<>();
+
+		if (verticalOffset == 0) {
+			return stickyLines;
+		}
+
+		try {
+			StyledText textWidget= sourceViewer.getTextWidget();
+			int startLine= textWidget.getTopIndex();
+
+			calculateStickyLinesForLineNumber(stickyLines, sourceViewer, startLine);
+			calculateStickyLinesUnderStickyLineControl(stickyLines, sourceViewer, startLine);
+		} catch (IllegalArgumentException e) {
+			stickyLines.clear();
+		}
+
+		return stickyLines;
+	}
+
+	private void calculateStickyLinesForLineNumber(LinkedList<StickyLine> stickyLines, ISourceViewer sourceViewer, int lineNumber) {
+		StyledText textWidget= sourceViewer.getTextWidget();
+		int startIndetation= getStartIndetation(lineNumber, textWidget);
+
+		for (int i= lineNumber, previousIndetation= startIndetation; i >= 0; i--) {
+			String line= textWidget.getLine(i);
+			int indentation= getIndentation(line);
+
+			if (indentation == IGNORE_INDENTATION) {
+				continue;
+			}
+
+			if (indentation < previousIndetation) {
+				previousIndetation= indentation;
+				stickyLines.addFirst(new StickyLine(line, mapLineNumberToSourceViewerLine(i, sourceViewer)));
+			}
+		}
+	}
+
+	private void calculateStickyLinesUnderStickyLineControl(LinkedList<StickyLine> stickyLines, ISourceViewer sourceViewer, int startLine) {
+		int firstBelowControl= startLine + stickyLines.size();
+		StyledText textWidget= sourceViewer.getTextWidget();
+		int lineCount= textWidget.getLineCount();
+
+		for (int i= startLine; i < firstBelowControl && i < lineCount; i++) {
```

Comment:
Is this a safe thing to do? For larger files, this can potentially become a problem.

Scrolling with Sticky Scrolling in a huge file is already not super smooth for me:
![14](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/109998bb-6cc0-4a70-a9d6-d7e2d5e033e7)
vs
![16](https://github.com/eclipse-platform/eclipse.platform.ui/assets/16443184/ea5c50f9-1d82-4035-a3a0-9764c89590f7)

A possible solution is to wait until scrolling is "done" to perform this analysis - and maybe do it in a separate thread (eclipse already has enough performance-issues in the UI-thread ;_;)

I don't believe this is too big of a deal since this feature can be opted out of easily

Reply:
The for loop does only check up to the maximum amount of sticky lines. So this should not be the bottle neck. The performance problem is caused by the fast scrolling  as the sticky lines are recalculated for every moved pixel. 

Waiting until the scrolling is done is not a proper solution, I think this will look buggy / laggy.

I will try to either move the calculation in another thread or skip some calculations when the editor is scrolled really fast.

Reply:
See https://github.com/eclipse-platform/eclipse.platform.ui/pull/1949